joomla/joomla-cms
# Shareable draft content feature

## Shareable draft content

This pull request adds a new feature called shareable draft content to Joomla. The goal 
of this feature is to provide the functionality of generating links that can be shared 
so users with this link can preview articles that have not yet been published.
### Implementation

The shareable links are implemented in several views of the content area:
- Articles list
- Article editing in the back-end
- Shared drafts view
- Article editing in the front-end
#### Articles list

The articles list has a new button called **Share draft**. This button generates one link per selected article.
#### Article editing in the back-end

The article editing page has a new button called **Share draft**. This button generates a link for the current article and
shows the generated link on the screen.
#### Shared drafts view

The shared drafts list lists all the created shared links. Here the shared links can be managed. The article titles link
to the article editing page if the user has permission to edit the given article.
#### Article editing in the front-end

The article editing page has a new button called **Share draft**. This button generates a link for the current article and
shows the generated link on the screen.
### Link generation

When a link is generated for the first time, a 16-character token will be created for that specific article. This token
is unique for each article and allows the user with the URL and token to access an unpublished article.

A token cannot be directly regenerated for an article as an existing token will always be returned. To create a new
token, the existing shared link needs to be deleted first from the Shared Drafts view.
#### Regular URL

By default a regular URL will be created.
#### SEF and Redirect Manager enabled

In case both SEF and the Redirect Manager are enabled, a SEF URL will be created so an easier to read URL can be used. The 
link is generated based on the alias of the article. Once a SEF URL is generated, this will be added to the Redirect
Manager table so it can be used to redirect the user to the correct article.
### Test instructions

The following instructions are based on regular URLs, make sure SEF is turned off.
#### Test single article link creation in back-end
1. Apply patch
2. Go to Extensions -> Manage -> Database
3. Click Fix (this is to create the database table)
4. Go to Content -> Articles
5. Create a new article
6. Notice there is **no** share draft button
7. Add a title and some content
8. Make sure the article is unpublished
9. Save the article
10. Notice **there is a** share draft button
11. Click on the Share draft button
12. A green box appears with the shareable link
13. Open the link and check that you can view the article
14. Close the article
#### Test multiple articles link creation
1. Select one or more articles
2. Click on the Share draft button
3. Notice there is a message saying that x links have been generated
4. Go to the Shared Drafts view en check that the articles and links are there
#### Check deletion of shareable links
1. Go to the Shared Drafts view
2. Select one or more links
3. Click the Delete button
4. Notice the message that one or more drafts have been deleted
#### Test single article link creation in front-end
1. Login to the front-end
2. Create a new article
3. Notice there is **no** share draft button
4. Login to the front-end
5. Edit an existing article
6. Notice **there is a** share draft button
7. Click on the Share draft button
8. A green box appears with the shareable link
9. Open the link and check that you can view the article
10. Close the article
#### SEF URLs
1. Login to the back-end
2. Go to System -> Global Configuration
3. Enable SEF URLs
4. Go to Components -> Redirects
5. Make sure that the Redirects plugin is enabled
6. Repeat the tests above and check that the generated URL is a SEF URL

Finally tagging @puneet0191 @chrisdavenport
